File: FunctionSampleDir/BlockTouchFragProject/app/src/main/java/com/mysample/retrotest/MainActivity.java
package com.mysample.retrotest;

import android.os.Handler;
import android.os.Message;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

/***
 * 프래그먼트를 엑티비티 위에 올려서 터치 이벤트 막기
 */

public class MainActivity extends AppCompatActivity
{
    long m_nReservedTime = 0;
    ReservedTimer m_hReservedHandler;
    TextView textView;
    ViewGroup MainContainer;

    FragmentTransaction transaction;
    WaitFragment fragment;
    RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        fragment = new WaitFragment();
        transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.MainContainer, fragment).commit();

        m_nReservedTime = System.currentTimeMillis() + 1000 * 30;
        m_hReservedHandler = new ReservedTimer();
        m_hReservedHandler.sendEmptyMessage(1 * 1000);

        initViewSetting();
    }

    private void initViewSetting()
    {
        MainContainer = findViewById(R.id.MainContainer);
        textView = findViewById(R.id.textView);
        textView.setOnClickListener(m_hClickListener);
        recyclerView = findViewById(R.id.recyclerView);
        Button button = findViewById(R.id.button);
        button.setOnClickListener(m_hClickListener);
    }

    View.OnClickListener m_hClickListener = new View.OnClickListener()
    {
        @Override
        public void onClick(View v) {
            switch (v.getId())
            {
                case R.id.button:
                    showToast("gg");
                    break;
                case R.id.textView:
                    showToast(textView.getText().toString());
                    break;
            }
        }
    };

    private void showToast(String strMessage)
    {
        Toast.makeText(this, strMessage, Toast.LENGTH_SHORT).show();
    }

    class ReservedTimer extends Handler
    {
        public void handleMessage(Message msg)
        {
            long nCur = System.currentTimeMillis();

            if(m_nReservedTime <= nCur)
            {
                clearReservedTime();
                return;
            }
            else
            {
                int nRemain = (int)(m_nReservedTime - nCur);
                String strText;

                nRemain /= 1000;

                strText = String.format("%d:%02d", nRemain/60, nRemain%60);
                textView.setText(strText);
                m_hReservedHandler.sendEmptyMessageDelayed(0, 1000);
            }
        }
    }

    private void clearReservedTime()
    {
        m_nReservedTime = 0;
        m_hReservedHandler.removeMessages(0);
        getSupportFragmentManager().beginTransaction().remove(getSupportFragmentManager().findFragmentById(R.id.MainContainer)).commit();
        MainContainer.setAlpha(1F);
    }
}

File: TestProject/EqualizerExoProject/app/src/main/java/com/mysample/exoeqtestproject/MainActivity.kt
package com.mysample.exoeqtestproject

import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.media.audiofx.AudioEffect
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.View
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*
import java.util.ArrayList

class MainActivity : AppCompatActivity(), View.OnClickListener
{
    companion object {
        val TESTURL = ""
    }

    val mPlayer : XwExoPlayer by lazy { XwExoPlayer(this) }
    val mEQFragment : EQFragment by lazy { EQFragment() }

    override fun onClick(v: View)
    {
        when(v.id)
        {
            R.id.btnPlay ->
            {
                showMessage("플레이")
                playRadio()
            }

            R.id.btnStop ->
            {
                stopRadio()
                showMessage("중지")
            }

            R.id.btnEQ ->
            {
//                val i = Intent()
//                i.putExtra(AudioEffect.EXTRA_PACKAGE_NAME, packageName)
//                i.putExtra(AudioEffect.EXTRA_AUDIO_SESSION, mPlayer.getSessionId())
//                startActivity(i)
                if(!mEQFragment.bViewState)
                {
//                    requestPermission()
                    attachEQFragment()
                }
                else
                {
                    detachEQFragment()
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initViewSetting()
    }

    private fun initViewSetting()
    {
        btnPlay.setOnClickListener(this)
        btnStop.setOnClickListener(this)
        btnEQ.setOnClickListener(this)
    }

    private fun showMessage(strMessage : String)
    {
        Toast.makeText(this, strMessage, Toast.LENGTH_SHORT).show()
    }

    private fun playRadio()
    {
        mPlayer.play(TESTURL)
        timeHandler.sendEmptyMessageDelayed(0,3 * 1000)
    }

    private fun stopRadio()
    {
        mPlayer.stop()
    }

    private var timeHandler: Handler = @SuppressLint("HandlerLeak") object :Handler()
    {
        override fun handleMessage(msg: Message)
        {
            if(mPlayer.getTitle().isNullOrEmpty())
            {
                sendEmptyMessageDelayed(0,3 * 1000)
            }
            else
            {
                tvTitle.text = mPlayer.getTitle()
            }
        }
    }

    fun attachEQFragment()
    {
        if(!mEQFragment.bViewState)
        {
            mEQFragment.setPlayer(mPlayer)
            val transaction = supportFragmentManager.beginTransaction()
            transaction.replace(R.id.llEQContainer, mEQFragment).commit()
        }
    }

    fun detachEQFragment()
    {
        if(mEQFragment.bViewState)
        {
            val transaction = supportFragmentManager.beginTransaction()
            transaction.remove(mEQFragment).commit()
        }
    }

    private val REQUEST_RESULT = 55
    val arr_permission = listOf(android.Manifest.permission.RECORD_AUDIO, android.Manifest.permission.MODIFY_AUDIO_SETTINGS)

    fun requestPermission()
    {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
        {
            val permissions = ArrayList<String>()
            for (permission in arr_permission)
            {
                var nPermissionCheck = checkSelfPermission(permission)
                if(nPermissionCheck != PackageManager.PERMISSION_GRANTED)
                {
                    permissions.add(permission)
                }
            }

            if(!permissions.isNullOrEmpty())
            {
                requestPermissions(permissions.toTypedArray(), REQUEST_RESULT)
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray)
    {
        if(requestCode == REQUEST_RESULT)
        {
            if(grantResults.isNotEmpty() && grantResults.contains(PackageManager.PERMISSION_GRANTED))
            {
                attachEQFragment()
            }
        }
    }
}
